### ISSUE TITLE
Unhandled exception when getting state

### ISSUE THREAD

**Issue by bachya** [bachya/aioguardian#40]
At random intervals while using this library in Home Assistant, I see this in the Home Assistant logs:

```txt
Traceback (most recent call last):
  File "/config/custom_components/guardian/__init__.py", line 225, in _async_update_listener_action
    await self.async_update()
  File "/config/custom_components/guardian/__init__.py", line 272, in async_update
    await asyncio.gather(*tasks)
  File "/config/custom_components/guardian/__init__.py", line 216, in _async_get_data_from_api
    result = await self._api_coros[api_category]()
  File "/usr/local/lib/python3.7/site-packages/aioguardian/commands/valve.py", line 105, in valve_status
    resp["data"]["state"] = VALVE_STATE_MAPPING[resp["data"]["state"]]
KeyError: 'state'
```

Super sporadic when it occurs.

**Comment by bachya** [comment]
Still diagnosing. When this occurs, it's because `resp` contains a `client.device.wifi_status()` response, _not_ a `client.valve.valve_status()` response. Running a new test; if that's the case, the best we can do is look for it and raise an exception (along with logging a bug to Elexa).

**Comment by bachya** [comment]
Tests aren't proving anything conclusive – this is still really sporadic. Going to handle the case and will submit a case to Elexa.

**Comment by bachya** [comment]
Over the past 12 hours, this has occurred a handful of times – no real pattern that I can see:

```txt
Error while fetching ping data: Sent command 0, but got response for command 80
Error while fetching sensor_status data: Sent command 80, but got response for command 16
Error while fetching valve_status data: Sent command 16, but got response for command 32
Error while fetching pair_sensor data: Sent command 48, but got response for command 0
Error while fetching diagnostics data: Sent command 1, but got response for command 48
Error while fetching valve_status data: Sent command 16, but got response for command 32
Error while fetching diagnostics data: Sent command 1, but got response for command 0
Error while fetching pair_sensor data: Sent command 48, but got response for command 80
Error while fetching ping data: Sent command 0, but got response for command 16
Error while fetching sensor_status data: Sent command 80, but got response for command 32
```

Interestingly, there is often an accompanying exception related to disconnection timeout (I think):

```txt

Error doing job: Task exception was never retrieved
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/aioguardian/client.py", line 87, in _execute_command
    data, remote_addr = await self._stream.recv()
  File "/usr/local/lib/python3.7/site-packages/asyncio_dgram/aio.py", line 103, in recv
    data, addr = await self._recvq.get()
  File "/usr/local/lib/python3.7/asyncio/queues.py", line 159, in get
    await getter
concurrent.futures._base.CancelledError

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/config/custom_components/guardian/__init__.py", line 225, in _async_update_listener_action
    await self.async_update()
  File "/config/custom_components/guardian/__init__.py", line 272, in async_update
    await asyncio.gather(*tasks)
  File "/config/custom_components/guardian/__init__.py", line 216, in _async_get_data_from_api
    result = await self._api_coros[api_category]()
  File "/usr/local/lib/python3.7/site-packages/aioguardian/commands/device.py", line 191, in wifi_status
    return await self._execute_command(Command.wifi_status, silent=silent)
  File "/usr/local/lib/python3.7/site-packages/aioguardian/client.py", line 89, in _execute_command
    raise SocketError(f"{command.name} command timed out")
  File "/usr/local/lib/python3.7/site-packages/async_timeout/__init__.py", line 55, in __aexit__
    self._do_exit(exc_type)
  File "/usr/local/lib/python3.7/site-packages/async_timeout/__init__.py", line 92, in _do_exit
    raise asyncio.TimeoutError
concurrent.futures._base.TimeoutError
```

I wonder if properly catching this will make any difference in the original issue... Will test.

**Comment by bachya** [comment]
Interesting: I looked at a graph of the device’s uptime and it appears to be rebooting at quasi-regular intervals.

![image](https://user-images.githubusercontent.com/47216/80929292-e61c8c80-8d67-11ea-8341-bc47375ec468.png)

**Comment by bachya** [comment]
More of the same over the last 18 hours:

```
Error while fetching valve_status data: Sent command 16, but got response for command 32
Error while fetching ping data: Sent command 0, but got response for command 16
Error while fetching diagnostics data: Sent command 1, but got response for command 0
Error while fetching pair_sensor data: Sent command 48, but got response for command 80
Error while fetching sensor_status data: Sent command 80, but got response for command 32
```

**Comment by bachya** [comment]
Well, fascinatingly, I deployed #43 about 3 hours ago and haven't had a single reoccurrence of this... Perhaps not handling timeouts correctly hindered the device's response? Will keep an eye out and close if nothing repeats further.

**Comment by bachya** [comment]
Showed up again, unfortunately – #43 wasn't the fix.

```txt
Error while fetching pair_sensor data: Sent pair_dump command, but got response for ping command
Error while fetching ping data: Sent ping command, but got response for sensor_status command
Error while fetching sensor_status data: Sent sensor_status command, but got response for valve_status command
Error while fetching valve_status data: Sent valve_status command, but got response for wifi_status command
```

**Comment by bachya** [comment]
Confirmed that the every-12-hours reboot is intended and nothing to worry about. 👍🏻

**Comment by bachya** [comment]
I spoke with a developer at Elexa about the incorrectly returned response payloads and he made two suggestions:

1. Since device communication happens over a single port, he recommended implementing a lock to ensure that only one command runs at a time.
2. Since the port is UDP, he recommended adding a "tens of milliseconds" delay with each API call to avoid drops.

Testing with (1) is very promising – in 12 hours, I saw no reoccurrences of the incorrect response payload issue.

I did notice 2-3 spots on those 12 hours where the commands timed out; going to experiment with (2) today.
